src/lib/Type.hs

```diff
@@ -115,27 +116,12 @@ getClassTy (ClassDef _ _ bs _ _) = go bs
 
 -- === querying effects ===
 
-computeAbsEffects :: (EnvExtender m, SubstE Name e)
-  => Abs (Nest Decl) e n -> m n (Abs (Nest Decl) (EffectRow `PairE` e) n)
-computeAbsEffects it = refreshAbs it \decls result -> do
-  effs <- declNestEffects decls
-  return $ Abs decls (effs `PairE` result)
-{-# INLINE computeAbsEffects #-}
+class HasEffectsE (e::E) where
+  effectsE :: (EnvReader m) => e n -> m n (EffectRow n)
```

Comment:
Couldn't we specialize this method to `EnvReaderM`? It's easy enough to `liftEnvReaderM`, or even to define
```
class HasEffectsE e where
  effectsEImpl :: e n -> EnvReaderM n (EffectRow n)

effectsE = liftEnvReaderM . effectsEImpl
{-# INLINE effectsE #-}
```
That would have the same ergonomics, but would let us guarantee that the relevant specializations are happening. At this point I tend to consider monad-polymorphic typeclass methods a smell (unless we're _very_ careful to inline them).

Reply:
Done.